### PR TITLE
BFT client pool: external clients parameter to replica configuration

### PR DESCRIPTION
**Pull request by Liorgal28**
In this PR I have added numOfExternalClients parameter to the replicaConfiguration in order to reserve pages to the external client in addition to the internal clients as well.

### CODE REVIEW

File: bftengine/include/bftengine/ReplicaConfig.hpp [hunk 1]
```diff
@@ -47,6 +47,11 @@ struct ReplicaConfig {
   // numOfClientProxies >= 1
   uint16_t numOfClientProxies = 0;
 
+  // number of objects that represent external clients.
+  // numOfExternalClients >= 0
+  // By default numOfExternalClients will be 0 unless we configure it will be greater than.
```
Comment (yuliasherman) [inline bftengine/include/bftengine/ReplicaConfig.hpp:52]: -> By default, numOfExternalClients is 0 unless configured differently.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 2]
```diff
@@ -2917,7 +2917,9 @@ ReplicaImp::ReplicaImp(bool firstTime,
   }
 
   std::set<NodeIdType> clientsSet;
-  for (uint16_t i = config_.numReplicas; i < config_.numReplicas + config_.numOfClientProxies; i++)
+  for (uint16_t i = config_.numReplicas;
+       i < config_.numReplicas + config_.numOfClientProxies + config_.numOfExternalClients;
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp:2921]: Don't re-calculate the loop end condition at every iteration: define a const variable before a loop and use it inside: 
const auto numOfEntities = config_.numReplicas + config_.numOfClientProxies + config_.numOfExternalClients;
for (uint16_t i = config_.numReplicas; I < numOfEntities....

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 3]
```diff
@@ -95,7 +95,8 @@ PreProcessor::PreProcessor(shared_ptr<MsgsCommunicator> &msgsCommunicator,
       maxReplyMsgSize_(myReplica.getReplicaConfig().maxReplyMessageSize - sizeof(ClientReplyMsgHeader)),
       idsOfPeerReplicas_(myReplica.getIdsOfPeerReplicas()),
       numOfReplicas_(myReplica.getReplicaConfig().numReplicas),
-      numOfClients_(myReplica.getReplicaConfig().numOfClientProxies),
+      numOfClients_(myReplica.getReplicaConfig().numOfClientProxies +
+                    myReplica.getReplicaConfig().numOfExternalClients),
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:99]: Preprocessor works only with external clients.

Comment (Liorgal28) [reply bftengine/src/preprocessor/PreProcessor.cpp:99]: so, remove numOfClientProxies? because before that he took only numOfClientProxies.

Comment (yuliasherman) [reply bftengine/src/preprocessor/PreProcessor.cpp:99]: Yes. Previously this number was expected to include all possible clients.

File: bftengine/include/bftengine/ReplicaConfig.hpp (hunk 1 above)
Comment (Liorgal28) [reply bftengine/include/bftengine/ReplicaConfig.hpp]: Fixed.

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 3 above)
Comment (Liorgal28) [reply bftengine/src/preprocessor/PreProcessor.cpp]: For now, we keep using the numOfClientProxies parameter in order not to break the Preprocessor tests.

**Pull request merged**